tls/s2n_handshake_io.c

```diff
@@ -1700,3 +1700,18 @@ int s2n_negotiate(struct s2n_connection *conn, s2n_blocked_status *blocked)
     conn->negotiate_in_use = false;
     return result;
 }
+
+S2N_RESULT s2n_fast_forward_handshake(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+    conn->handshake.handshake_type = NEGOTIATED | FULL_HANDSHAKE;
```

Comment:
It's complicated. So the motivation for needing to skip over the handshake in deserialization actually comes from [this line](https://github.com/aws/s2n-tls/blob/ba825b878a58079e7453fe3c6f00b48a8752398c/tls/s2n_tls13_secrets.c#L698) when you're generating the keys for TLS1.3. To generate the master secret the handshake must be complete, which means that the state machine must point to the [APPLICATION_DATA entry](https://github.com/aws/s2n-tls/blob/ba825b878a58079e7453fe3c6f00b48a8752398c/tls/s2n_handshake_io.c#L1594).
To get to that entry, you have to 1. pick a state machine, 2. pick a handshake type that includes the APPLICATION_DATA entry (some handshakes do not contain application data), and then 3. iterate through that handshake type until you find the APPLICATION_DATA entry.
So that's the whole rational for setting a handshake_type here.

Reply:
I think it's unexpected just based on what the method is supposed to do. You might want to set the handshake type outside of the reusable method so that it doesn't always set the handshake type. I wonder if it's also worth resetting it once you're done, since we don't really have a handshake type? Or we could put the handshake type in the serialized state, but that really doesn't seem necessary. We could also consider a handshake specifically for this serialization case, but that's also probably overkill.

Reply:
I dunno, we've written our code with the perspective of always having a handshake type set. Like, I wonder if it's actually safer to keep a handshake type set, even if that may not be the actually negotiated type. I kind of see both sides to this.

Reply:
Yeah currently s2n_shutdown uses the handshake type to determine if the handshake is complete:
https://github.com/aws/s2n-tls/blob/3926a458aafdeb4c63ddefa63e2c54b431407e92/tls/s2n_shutdown.c#L43-L45

Maybe `s2n_handshake_is_complete` could use the flag that Lindsay mentioned in https://github.com/aws/s2n-tls/pull/4468#discussion_r1544754416? I'm not sure if there are other places where the handshake type is used after the handshake though.